### PR TITLE
Solutions Feature Request Report

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

A new dropdown option is added called "Solutions Feature Request":
Screenshot from 2024-01-24 15-20-30

Clicking on this button will cause a popup form to show:
Screenshot from 2024-01-24 15-20-43

After clicking Submit, a Jira ticket will be created in [this](https://dimagi-dev.atlassian.net/jira/servicedesk/projects/SF/queues/custom/185) project space.

It should be noted that this new Solutions feature request form is **only available to staff members**. Users that are not staff will not see the new dropdown option or the new modal form.

## Technical Summary

Link to ticket [here](https://dimagi-dev.atlassian.net/browse/SC-3317).

This PR adds a reporting system very similar to bug reports where a user can fill out a form which will automatically send an email. This email will, in turn, automatically create a Jira issue.

~~**Note:** This PR should only be merged once the changes from [this PR](https://github.com/dimagi/commcare-cloud/pull/6211) have been merged.~~ The linked PR will no longer be merged and the variable for the Solutions feedback email address will be inserted into CommCare HQ's code with this PR.

## Feature Flag

None

### Safety story

- Local testing
- Unit test
- QA passed

### Automated test coverage

Unit test exist for view sending email.

### QA Plan

QA ticket has been created and is available [here](https://dimagi.atlassian.net/browse/QA-6062).

Note: QA has passed.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html
Comment (mkangia) [inline corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html:85]: nit: how about "file containing sensitive data like passwords or project data"

I guess its important to be explicit to user what should be avoided.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap5/modal_solutions_feature_request.html
Comment (mkangia) [inline corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap5/modal_solutions_feature_request.html:74]: how is the format validated for emails?

File: corehq/apps/hqwebapp/templates/hqwebapp/base.html
Comment (mkangia) [inline corehq/apps/hqwebapp/templates/hqwebapp/base.html:216]: just confirming that this would not fail if there is no `couch_user` on request? like a non-logged in page.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/global_navigation_bar.html
Comment (mkangia) [inline corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/global_navigation_bar.html:132]: just confirming that this would not fail if there is no `user`? like a non-logged in page.

File: corehq/apps/hqwebapp/views.py
Comment (mkangia) [inline corehq/apps/hqwebapp/views.py:836]: this should support both right? report_issue and feature_request?

File: corehq/apps/hqwebapp/views.py
Comment (mkangia) [inline corehq/apps/hqwebapp/views.py:821]: nit: this is redundant now and can be removed since the ICDS environment does not exist anymore. Please feel to leave if don't want to make this change in this PR since its unrelated

File: corehq/apps/hqwebapp/views.py
Comment (mkangia) [inline corehq/apps/hqwebapp/views.py:865]: lets also check here that the user is a staff member?

File: corehq/apps/hqwebapp/views.py
Comment (mkangia) [inline corehq/apps/hqwebapp/views.py:869]: that might have been a better name actually.
https://github.com/dimagi/commcare-cloud/pull/6211#discussion_r1465922233

File: corehq/apps/hqwebapp/views.py
Comment (SmittieC) [inline corehq/apps/hqwebapp/views.py:836]: Will this not break the `report_issue` flow?

**Review by mkangia — CHANGES_REQUESTED** [review]
Hey Zandre,

Great PR, good flow of commits.
Left some blocking feedback to consider for improvements.

File: corehq/apps/hqwebapp/views.py
Comment (SmittieC) [inline corehq/apps/hqwebapp/views.py:871]: Might be good to tell the user all went well?

File: corehq/apps/hqwebapp/views.py
Comment (zandre-eng) [reply corehq/apps/hqwebapp/views.py:836]: Ah, good catch! You're right that this should not be changed, I must have updated this by accident. Will revert it back.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html:85]: Sounds good, it would make it a bit more clear to the user. Since I mostly copied this template from the bug report, I suppose we should change the text there as well to be more descriptive?

Comment (mkangia) [reply corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html:85]: Good to suggest that to SaaS first. I don't think we should simply change that modal though this one should be safe to update.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap5/modal_solutions_feature_request.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap5/modal_solutions_feature_request.html:74]: I have added the missing JS files. The validation is happening in these files.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_solutions_feature_request.html]: Good point. Given that, I'll leave the bug report text for now as it's out of scope for this PR.

File: corehq/apps/hqwebapp/templates/hqwebapp/base.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/base.html:216]: Since the button that shows this modal is only viewable for logged in users, there should always be a `couch_user`.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/global_navigation_bar.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/global_navigation_bar.html:132]: Please refer to [this](https://github.com/dimagi/commcare-hq/pull/34026#discussion_r1466121235) comment.

File: corehq/apps/hqwebapp/views.py
Comment (zandre-eng) [reply corehq/apps/hqwebapp/views.py]: The `BugReportView` should only support `report_issue` since that is from the modal it uses. Likewise, `feature_request` should only be in `SolutionsFeatureRequestView`

File: corehq/apps/hqwebapp/views.py
Comment (zandre-eng) [reply corehq/apps/hqwebapp/views.py:821]: I think given that I simply moved what was being done for the SaaS bug report, I'll opt to rather leave this as is for now.

File: corehq/apps/hqwebapp/views.py
Comment (zandre-eng) [reply corehq/apps/hqwebapp/views.py:871]: After adding the missing JS files ([1b19cdf](https://github.com/dimagi/commcare-hq/pull/34026/commits/1b19cdf9ab50b6d12a666e4b8256751a0cd6b3ab)), the user should now see a success message as follows:
Screenshot from 2024-01-25 12-23-39

File: corehq/apps/hqwebapp/templates/hqwebapp/base.html
Comment (mkangia) [reply corehq/apps/hqwebapp/templates/hqwebapp/base.html:216]: okay, I assume the modal template will always be there on the page, irrespective of its usage since its in the base template.
Can you confirm that this was tested for a non logged in user?

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/solutions-feature-request.js
Comment (mkangia) [inline corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/solutions-feature-request.js:1]: So this is the one same as for reporting an issue.

File: corehq/apps/hqwebapp/views.py
Comment (mkangia) [inline corehq/apps/hqwebapp/views.py:865]: should we return a 400 or similar in case the user isn't a staff?

File: corehq/apps/hqwebapp/templates/hqwebapp/base.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/base.html:216]: I can confirm this works for a non-logged in user as well. The bug report uses a similar approach with the `{% if allow_report_an_issue %}` where `allow_report_an_issue` is a permission check on the couch user.

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/solutions-feature-request.js
Comment (zandre-eng) [reply corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/solutions-feature-request.js:1]: This one was similar in that it was mostly copied from the JS file for reporting an issue. I have since changed this though so that both modals now use the same generic JS file using knockout.

File: corehq/apps/hqwebapp/views.py
Comment (zandre-eng) [reply corehq/apps/hqwebapp/views.py:865]: Good idea, this would make it more obvious that something went wrong.

Comment (zandre-eng) [reply corehq/apps/hqwebapp/views.py]: [df1e692](https://github.com/dimagi/commcare-hq/pull/34026/commits/df1e6928086376179dab184076e138ab9612fb9b)

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js
Comment (mkangia) [inline corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js:109]: what is "success" here?

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_report_issue.html
Comment (mkangia) [inline corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_report_issue.html:28]: just confirming that these classes are added if `hasSubjectError` is `true`?

**Review by mkangia — COMMENTED** [review]
Thanks for the follow up @zandre-eng and the effort to extract out a common email-request JS file.

Couple of things to confirm:
1. How can we confirm that there is no change to how the bug report works today? Since we have changed the form to be handled by knockout now, things are being done a bit differently now.
2. We are to change the name of the "internal feedback email" to solution specific one and add dimagi environment check?

Additionally, good to highlight in the description that this is only available to staff members otherwise it appears that its available to all users.

Changes look good. Just holding off on approval till we are through QA to confirm changes functionally.

**Comment by zandre-eng** [comment]
> I see there's different templates for bootstrap 3 and 5; do you know where the correct template is decided upon? I don't think I'm 100% sure how this works.

@Charl1996 This gets determined in `corehq/apps/hqwebapp/templates/hqwebapp/base.html`, specifically with the `{% if use_bootstrap5 %}` check. Depending on the result of this check, the correct template version will be loaded.

File: corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_report_issue.html
Comment (zandre-eng) [reply corehq/apps/hqwebapp/templates/hqwebapp/includes/bootstrap3/modal_report_issue.html:28]: That is correct. You can also take a look at the relevant [knockoutJS documentation](https://knockoutjs.com/documentation/css-binding.html) which shows an example of using a true/false variable in this manner.

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js
Comment (zandre-eng) [reply corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js:109]: This will put the form submit button in the success state. Specifically, it will cause the text from the `data-success-text` binding to be shown.

**Comment by zandre-eng** [comment]
> Thanks for the follow up @zandre-eng and the effort to extract out a common email-request JS file.

@mkangia Thank you for the review. Regarding your questions:
1. Although things are slightly different now, I was quite meticulous at ensuring I simply replace the current logic with a relevant KO binding (and therefore not change the old logic). In saying that however, this will be run through QA to ensure that the bug inquiry form still functions correctly.
2. Please refer to [08380db](https://github.com/dimagi/commcare-hq/pull/34026/commits/08380db86c26f2941a7647b6b4d362004a7ace25) commit.

That is a good point, I have made the necessary description changes to better reflect this.

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js
Comment (zandre-eng) [reply corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js]: This was simply copied over from the old bug inquiry logic, although I do agree it is a bit weird to be comparing translated strings. I have updated this to use a more straightforward `isReportSent` flag ([2d377ef](https://github.com/dimagi/commcare-hq/pull/34026/commits/2d377efe8c4dea65896c13dd2c836e895b9ce825))

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js
Comment (mkangia) [inline corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap3/email-request.js:57]: :100:

**Comment by mkangia** [comment]
Thanks for the follow up @zandre-eng 

Cool, will look forward to QA before I set this as approved. No blocking feedback now.

> only available to Dimagi staff members

One small detail here would be to simply say "only available to Staff members" or "users marked as staff" or "users with staff access" to keep this generic.

File: corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap5/email-request.js
Comment (mkangia) [inline corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap5/email-request.js:39]: should the name be changed here as well? `isSubjectEmpty`.

Comment (zandre-eng) [reply corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap5/email-request.js:39]: Ah, yes that should. This should match how we have it for the bootstrap 3 version.

Comment (zandre-eng) [reply corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap5/email-request.js]: [d0ebdda](https://github.com/dimagi/commcare-hq/pull/34026/commits/d0ebddaffb18f44e2f954488a3abcb2fa6772563)

File: settings.py
Comment (mkangia) [inline settings.py:494]: we don't need this now yes?

**Review by mkangia — APPROVED** [review]

File: settings.py
Comment (zandre-eng) [reply settings.py]: We don't, I'll remove this ([1d22c49](https://github.com/dimagi/commcare-hq/pull/34026/commits/1d22c49ab148ea3cb98b77a1adadc10bcf9fea75))

**Review by mkangia — APPROVED** [review]

**Review by Charl1996 — APPROVED** [review]
Nice work, @zandre-eng !

**Comment by orangejenny** [comment]
FYI, this broke the report an issue functionality for B5 pages, and I don't think the solutions feature request modal has ever worked on B5 pages. There's logic around validating the subject that's missing from the B5 templates (see [this diff line](https://github.com/dimagi/commcare-hq/pull/34026/files#diff-dae80f3e0284df244b70a8e926ed2b3d4fe7877b2fbc47ff95a6dbbc2227051dR78) for some code that I suspect should be in both B3 and B5 versions of the template), which is causing validation to always fail, without indicating why to the user.

**Comment by zandre-eng** [comment]
> FYI, this broke the report an issue functionality for B5 pages, and I don't think the solutions feature request modal has ever worked on B5 pages. There's logic around validating the subject that's missing from the B5 templates (see [this diff line](https://github.com/dimagi/commcare-hq/pull/34026/files#diff-dae80f3e0284df244b70a8e926ed2b3d4fe7877b2fbc47ff95a6dbbc2227051dR78) for some code that I suspect should be in both B3 and B5 versions of the template), which is causing validation to always fail.

@orangejenny Thanks for flagging this, I'll take a look into it. I think all the testing was done on B3 pages which is probably how this managed to slip past.

**Comment by orangejenny** [comment]
@zandre-eng Cool, thank you. Yeah, I was looking at the date and figuring there might not have even been any B5 pages to test on at the time.

**Pull request merged**
